presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java

```diff
@@ -0,0 +1,575 @@
+    {
+        boolean leftStrictlyNegative = isStrictlyNegative(leftRawLo, leftRawHi);
+        boolean rightNegative = isNegative(rightRawLo, rightRawHi);
+
+        if (leftStrictlyNegative != rightNegative) {
+            return leftStrictlyNegative ? -1 : 1;
+        }
+        else {
+            return compareUnsigned(leftRawLo, leftRawHi, rightRawLo, rightRawHi) * (leftStrictlyNegative ? -1 : 1);
+        }
+    }
+
+    public static int compareUnsigned(Slice left, Slice right)
+    {
+        long leftHi = getLong(left, 1);
+        long rightHi = getLong(right, 1);
+        if (leftHi != rightHi) {
+            return Long.compareUnsigned(leftHi, rightHi);
+        }
+
+        long leftLo = getLong(left, 0);
+        long rightLo = getLong(right, 0);
+        if (leftLo != rightLo) {
+            return Long.compareUnsigned(leftLo, rightLo);
+        }
+
+        return 0;
+    }
+
+    public static int compareUnsigned(long leftRawLo, long leftRawHi, long rightRawLo, long rightRawHi)
+    {
+        long leftHi = unpackUnsignedLong(leftRawHi);
+        long rightHi = unpackUnsignedLong(rightRawHi);
+        if (leftHi != rightHi) {
+            return Long.compareUnsigned(leftHi, rightHi);
+        }
+
+        if (leftRawLo != rightRawLo) {
+            return Long.compareUnsigned(leftRawLo, rightRawLo);
+        }
+
+        return 0;
+    }
+
+    public static void negate(Slice decimal)
+    {
+        setNegative(decimal, !isNegative(decimal));
+    }
+
+    public static boolean isStrictlyNegative(Slice decimal)
+    {
+        return isNegative(decimal) && (getLong(decimal, 0) != 0 || getLong(decimal, 1) != 0);
+    }
+
+    public static boolean isStrictlyNegative(long rawLo, long rawHi)
+    {
+        return isNegative(rawLo, rawHi) && (rawLo != 0 || unpackUnsignedLong(rawHi) != 0);
+    }
+
+    public static boolean isNegative(Slice decimal)
+    {
+        return (getRawInt(decimal, SIGN_INT_INDEX) & SIGN_INT_MASK) != 0;
+    }
+
+    public static boolean isNegative(long rawLo, long rawHi)
+    {
+        return (rawHi & SIGN_LONG_MASK) != 0;
+    }
+
+    public static long hash(Slice decimal)
+    {
+        return hash(getRawLong(decimal, 0), getRawLong(decimal, 1));
+    }
+
+    public static long hash(long rawLo, long rawHi)
+    {
+        return XxHash64.hash(rawLo) ^ XxHash64.hash(unpackUnsignedLong(rawHi));
+    }
+
+    private static void scaleDownFive(Slice decimal, int fiveScale, Slice result)
+    {
+        while (true) {
+            int powerFive = Math.min(fiveScale, MAX_POWER_FIVE_INT);
+            fiveScale -= powerFive;
+
+            int divisor = POWER_FIVES_INT[powerFive];
+            divide(decimal, divisor, result);
+            decimal = result;
+
+            if (fiveScale == 0) {
+                return;
+            }
+        }
+    }
+
+    private static void shiftRightRoundUp(Slice decimal, int rightShifts, Slice result)
```

Comment:
You can imagine having accumulator to which you would write result. Such accumulator would be returned as operator result. This is probably next logical step in optimizing performance (see ArrayConcatFunction how such accumulator can be used)